bakery/macaroon.go

```diff
@@ -0,0 +1,352 @@
+package bakery
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+
+	"golang.org/x/net/context"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/macaroon.v2-unstable"
+
+	"gopkg.in/macaroon-bakery.v2-unstable/bakery/checkers"
+)
+
+// legacyNamespace holds the standard namespace as used by
+// pre-version3 macaroons.
+func legacyNamespace() *checkers.Namespace {
+	ns := checkers.NewNamespace(nil)
+	ns.Register(checkers.StdNamespace, "")
+	return ns
+}
+
+// Macaroon represents an undischarged macaroon along its first
+// party caveat namespace and associated third party caveat information
+// which should be passed to the third party when discharging a caveat.
+type Macaroon struct {
+	// m holds the underlying macaroon.
+	m *macaroon.Macaroon
+
+	// version holds the version of the macaroon.
+	version Version
+
+	// caveatData maps from a third party caveat id to its
+	// associated information, usually public-key encrypted with the
+	// third party's public key.
+	//
+	// If version is less than Version3, this will always be nil,
+	// because clients prior to that version do not support
+	// macaroon-external caveat ids.
+	caveatData map[string][]byte
+
+	// namespace holds the first-party caveat namespace of the macaroon.
+	namespace *checkers.Namespace
+
+	// caveatIdPrefix holds the prefix to use for the ids of any third
+	// party caveats created. This can be set when Discharge creates a
+	// discharge macaroon.
+	caveatIdPrefix []byte
+}
+
+type macaroonJSON struct {
+	Macaroon *macaroon.Macaroon `json:"m"`
+	Version  Version            `json:"v"`
+	// Note: CaveatData is encoded using URL-base64-encoded keys
+	// because JSON cannot deal with arbitrary byte sequences
+	// in its strings, and URL-base64 values to match the
+	// standard macaroon encoding.
+	CaveatData map[string]string   `json:"cdata,omitempty"`
+	Namespace  *checkers.Namespace `json:"ns"`
+}
+
+// Clone returns a copy of the macaroon. It does not
```

Comment:
This second sentence is a little ambiguous. I read it first as the resulting macaroon will have no namespace, consider rewording it.